pkg/planner/cascades/memo/group.go

```diff
@@ -198,16 +218,97 @@ func (g *Group) ForEachGE(f func(ge *GroupExpression) bool) {
 
 // removeParentGEs remove the current Group's parent GE ref which is pointed to parent.
 func (g *Group) removeParentGEs(parent *GroupExpression) {
-	_, ok := g.hash2ParentGroupExpr.Get(parent)
+	addr := uintptr(unsafe.Pointer(parent))
+	_, ok := g.hash2ParentGroupExpr.Get(addr)
 	intest.Assert(ok)
-	g.hash2ParentGroupExpr.Remove(parent)
+	g.hash2ParentGroupExpr.Remove(addr)
 }
 
 // addParentGEs is used to maintain the reverted parent pointer from Group to parent GEs.
 func (g *Group) addParentGEs(parent *GroupExpression) {
-	_, ok := g.hash2ParentGroupExpr.Get(parent)
+	addr := uintptr(unsafe.Pointer(parent))
+	_, ok := g.hash2ParentGroupExpr.Get(addr)
 	intest.Assert(!ok)
-	g.hash2ParentGroupExpr.Put(parent, struct{}{})
+	g.hash2ParentGroupExpr.Put(addr, parent)
+}
+
+// mergeTo will merge src group's element into target group.
+func (g *Group) mergeTo(target *Group) {
+	// maintain target group's parent GE refs, except the triggering src GE.
+	g.hash2ParentGroupExpr.Each(func(key uintptr, val *GroupExpression) {
+		target.hash2ParentGroupExpr.Put(key, val)
+	})
+	// maintain the ge migration, two groups may have the equivalent two,
+	// check the duplication when inserting, merge their GE's state together.
+	g.ForEachGE(func(ge *GroupExpression) bool {
+		existedElem, ok := target.hash2GroupExpr.Get(ge)
+		if ok {
+			existedGE := existedElem.Value.(*GroupExpression)
+			// not a same object
+			intest.Assert(ge != existedGE)
+			ge.mergeTo(existedGE)
+		} else {
+			// change the owner group inside.
+			// since ge's is migrated from src group, the underlying parentGE
+			// ref from child groups is not necessary to change.
+			target.Insert(ge)
+		}
+		// note:
+		// 1: for distinct ge among two groups: since ge's global register info, mm.hash2GlobalGroupExpr has already done
+		// when inserting themselves into src group or dst group, nothing to do here. childG's parentGERefs is also maintained
+		// before, nothing to do here.
+		// 2: for the duplicated ge in two groups, which is the root cause for a recursive group merge of this two groups, as
+		// the same way, mm.hash2GlobalGroupExpr is done once before, the re-inserted ge doesn't do it again, so nothing to do
+		// here. childG's parentGERefs is done for re-inserted GE, but it's cleared ge.MergeTo(existedGE) here.
+		return true
+	})
+	g.Clear()
+}
+
+// Clear clean and release the group element in the container.
+func (g *Group) Clear() {
+	// clean the list.
+	g.hash2GroupExpr.Each(func(key *GroupExpression, val *list.Element) {
```

Comment:
We can observe the memory alloc here, which may be a point where memory can be reused, but the optimization here can be discussed later.

Reply:
changed as 
`g.logicalExpressions.Init()` which is the list.clear way